src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -758,102 +766,84 @@ protected virtual void CheckDocumentCanBeRemoved(DocumentId documentId)
         /// </summary>
         protected internal void OnDocumentTextLoaderChanged(DocumentId documentId, TextLoader loader)
         {
-            using (_serializationLock.DisposableWait())
-            {
-                CheckDocumentIsInCurrentSolution(documentId);
-
-                var oldSolution = this.CurrentSolution;
-
-                var newSolution = oldSolution.WithDocumentTextLoader(documentId, loader, PreservationMode.PreserveValue);
-                newSolution = this.SetCurrentSolution(newSolution);
-
-                var newDocument = newSolution.GetDocument(documentId)!;
-                this.OnDocumentTextChanged(newDocument);
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentChanged, oldSolution, newSolution, documentId: documentId);
-            }
+            SetCurrentSolution(
+                oldSolution =>
+                {
+                    CheckDocumentIsInSolution(oldSolution, documentId);
+                    return oldSolution.WithDocumentTextLoader(documentId, loader, PreservationMode.PreserveValue);
+                },
+                WorkspaceChangeKind.DocumentChanged,
+                onAfterUpdate: (_, newSolution) => this.OnDocumentTextChanged(newSolution.GetRequiredDocument(documentId)),
+                documentId: documentId);
```

Comment:
here we want to issue the normal event, and a special 'doc changed' event.  so we override onAfterUpdate here.